Implement constructors directly from datasets for reference objects

CPBridge opened the pull request:
In the SR module there are a number of classes that represent various types of references to source images:

- highdicom.sr.content.ReferencedSegment
- highdicom.sr.content.ReferencedSegmentationFrame
- highdicom.sr.content.SourceImageForMeasurement
- highdicom.sr.content.SourceImageForRegion
- highdicom.sr.content.SourceImageForSegmentation
- highdicom.sr.content.SourceSeriesForSegmentation

The constructors for these classes require that the user manually pass the relevant UIDs. This is by no means a large burden but is a little low-level and increases the probability of a mistake.

I propose implementing alternative constructors as class methods called `from_dataset` that allow constructing objects by simply passing the relevant dataset and have highdicom take care of the rest.

For example

```python
import pydicom
from highdicom.sr.content import SourceImageForMeasurement

ct_dataset = pydicom.dcmread(...)

# Currently a source image object must be constructed like this
source_image = SourceImageForMeasurement(
    referenced_sop_class_uid=ct_dataset.SOPClassUID,
    referenced_sop_instance_uid=ct_dataset.SOPInstanceUID
)

# Under the proposal, it could also alternatively be constructed like this
source_image = SourceImageForMeasurement.from_dataset(ct_dataset)
```

I personally think this is a bit higher level and a nice reduction in boilerplate.

For now, I have just implemented this for the `SourceImageForMeasurement` class. If you like the idea, I will implement for the other classes and write tests.

hackermd commented:
@CPBridge I like the idea!

What do you think about calling the classmethod `from_source()` instead of `from_dataset()`? The constructor of these classes will also create a `Dataset`, therefore this could get confusing.

CPBridge commented:
Cool, I'll get to work on the rest then. I agree that dataset is a little confusing as all these objects are datasets. I could get behind `from_source()`. Or maybe the slightly more verbose `from_source_dataset()`, since `source` is also a bit vague?

CPBridge commented:
@hackermd this is now implemented and ready for review. The referenced segment and referenced segmentation frame versions were more complex than I thought since there are a few different ways that source information may appear in a segmentation dataset. But this also means it will be more useful. Hopefully I've covered most eventualities.

CPBridge commented:
@hackermd this is awaiting rewiew

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -103,6 +105,34 @@ def __init__(
         )
 
+    @classmethod
+    def from_source_dataset(
```
Given that we are considering adding a `from_dataset()`, it may be better to call this method just `from_source_image()`. What do you think?

P.S.: If we do this, we should apply the same logic to the other methods as well.

CPBridge replied:
Yeah, agreed. Changed for all relevant classes

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+            # Use only the provided frames
+            for f in frame_numbers:
+                # Check that the provided frame number is valid
+                if f < 0 or f >= dataset.NumberOfFrames:
```
Frame numbers are one-based.

```suggestion
                if f < 1 or f >= dataset.NumberOfFrames:
```

CPBridge replied:
Addressed in 4bea49d

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+                        'provided dataset'
+                    )
+
+                frame_info = dataset.PerFrameFunctionalGroupsSequence[f]
```
```suggestion
                i = f - 1
                frame_info = dataset.PerFrameFunctionalGroupsSequence[i]
```

CPBridge replied:
Addressed in 4bea49d

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+                referenced_frame_info.append(frame_info)
+        else:
+            # Use every frame that applies to this segment
+            for frame_info in dataset.PerFrameFunctionalGroupsSequence:
```
Since there may be many frames in a Segmentation instance, I would rewrite this into a list comprehension for improved performance.

CPBridge replied:
Addressed in 4bea49d

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+        # Attempt to retrieve a list of source images for the specific segment
+        # and frames specified using information in the per-frame functional
+        # groups sequence
+        # This gives more specific information, but depends upon optional
```
Which attributes are optional? Based on my understanding, all the used attributes are required. Am I missing something?

hackermd replied:
Ah...this seems to not only refer to the next block, but also to the subsequent block.

CPBridge replied:
Yeah exactly. Have clarified the comments somewhat in  4bea49d

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+        # referenced frames
+        refd_insuids = set()
+        for frame_info in referenced_frame_info:
+            if hasattr(frame_info, 'DerivationImageSequence'):
```
The two lines could be combined into one:

```python
for drv_image in getattr(frame_info, 'DerivationImageSequence', []):
    ...
```

CPBridge replied:
Neat. Addressed in 4bea49d

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+            if hasattr(frame_info, 'DerivationImageSequence'):
+                for drv_image in frame_info.DerivationImageSequence:
+                    if hasattr(drv_image, 'SourceImageSequence'):
+                        for src_image in drv_image.SourceImageSequence:
```
The two lines could be combined into one:

```python
for src_image in getattr(drv_image, 'SourceImageSequence', []):
    ...
```

CPBridge replied:
Addressed in 4bea49d

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+                            ins_uid = src_image.ReferencedSOPInstanceUID
+                            cls_uid = src_image.ReferencedSOPClassUID
+                            if ins_uid not in refd_insuids:
+                                if hasattr(src_image, 'ReferencedFrameNumber'):
+                                    ref_frames = src_image.ReferencedFrameNumber
+                                else:
+                                    ref_frames = None
```
```suggestion
                                ref_frames = getattr(src_image, 'ReferencedFrameNumber', None)
```

CPBridge replied:
Addressed in 4bea49d

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+                'Input dataset should be a segmentation storage instance'
+            )
+
+        source_images = []
```
Why not `source_images = None`?

CPBridge replied:
Because we need to append to it later in the loop beginning:

```python
for frame_info in referenced_frame_info:
```
currently on line 1006

hackermd replied:
I see. It's potentially error prone due to the `source_images if source_images else None`, but I don't have a better suggestion either.

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+                    source_series = SourceSeriesForSegmentation(
+                        ref_series.SeriesInstanceUID
+                    )
+                    source_images = []
```
Is this necessary? This does not (and is not supposed to) change the value assigned to the variable, does it?

CPBridge replied:
Nope, should be redundant. Removed in 4bea49d

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+            sop_instance_uid=dataset.SOPInstanceUID,
+            segment_number=segment_number,
+            frame_numbers=frame_numbers,
+            source_images=source_images if source_images else None,
```
This may not be necessary. See above comments.

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+                    )
+            else:
+                raise AttributeError(
+                    "Dataset does not contain a ReferencedSeriesSequence "
```
I would clarify that the **segmentation** dataset does not contain the attribute (similar to the above error message):

```suggestion
                    "Segmentation dataset does not contain a ReferencedSeriesSequence "
```

CPBridge replied:
Addressed in 4bea49d

hackermd commented on tests/test_sr.py:
```diff
@@ -819,6 +821,660 @@ def test_finding_site(self):
+        self._src_dataset = dcmread(
+            str(data_dir.joinpath('test_files', 'ct_image.dcm'))
+        )
+        self._ref_frames = [0, 4, 5]
```
Frame numbers shall start at 1:

```suggestion
        self._ref_frames = [1, 4, 5]
```

hackermd replied:
Applies to the other test cases as well.

CPBridge replied:
Addressed in 4bea49d. Also added a test on frame number (previously there was no check)

hackermd commented on tests/test_sr.py:
```diff
@@ -819,6 +821,660 @@ def test_finding_site(self):
+        )
+        self._ref_frames = [0, 4, 5]
+
+    def test_basic(self):
```
I would suggest naming the method `test_construction` instead for clarity.

Applies to the other test cases as well.

CPBridge replied:
Addressed in 4bea49d

hackermd commented on tests/test_sr.py:
```diff
@@ -819,6 +821,660 @@ def test_finding_site(self):
+            self._src_dataset.SOPInstanceUID
+        )
+
+    def test_basic_frames(self):
```
I would suggest naming the method `test_construction_with_frame_reference` instead for clarity.

CPBridge replied:
Addressed in 4bea49d

CPBridge commented on src/highdicom/sr/content.py:
```diff
@@ -169,6 +170,34 @@ def __init__(
+
+        """
+        return cls(
+            referenced_sop_class_uid=dataset.SOPClassUID,
```
Is there a well-defined list of SOP Classes that would be acceptable here? Seems like we don't want to be overly restrictive. Or would we instead create a blacklist containing obvious mistakes such as Segmentation, SRs, GSPS etc?

hackermd replied:
If it walks like a duck and talks like a duck...

CPBridge replied:
...then we can say with a high degree of certainty that (0010,2201) contains "Anatidae"?

hackermd replied:
[PS3.3 A.1.2.6 Image IE](http://dicom.nema.org/medical/dicom/current/output/html/part03.html#sect_A.1.2.6)

> The Image IE defines the Attributes that describe the pixel data of an image. 

However, that definition includes Segmentation images.

hackermd replied:
> Or would we instead create a blacklist containing obvious mistakes such as Segmentation, SRs, GSPS etc?

Maybe we should just assert that the `PixelData` attribute (and its friends for floating point values) is present and that `Modality` is not obviously wrong (`SR`, `SEG`, etc.) without being overly restrictive or comprehensive.

CPBridge commented on src/highdicom/sr/content.py:
```diff
@@ -523,6 +655,113 @@ def __init__(
+        -----
+             This method will attempt to deduce source image information
+             from information provided in the segmentation instance. If
+             available, it will used information specific to the segment
```
Adressed in 4bea49d

CPBridge commented on src/highdicom/sr/content.py:
```diff
@@ -523,6 +655,113 @@ def __init__(
+             from information provided in the segmentation instance. If
+             available, it will used information specific to the segment
+             and frame numbers (if any) provided using the Derivation
+             Image Sequence information in the frames. If this information
```
Addressed in 4bea49d

CPBridge commented on src/highdicom/sr/content.py:
```diff
@@ -523,6 +655,113 @@ def __init__(
+
+        frame_info = dataset.PerFrameFunctionalGroupsSequence[frame_number]
+
+        segment_number = frame_info.SegmentIdentificationSequence[0]\
```
Addressed in 4bea49d

CPBridge commented on src/highdicom/sr/content.py:
```diff
@@ -523,6 +655,113 @@ def __init__(
+                'provided dataset'
+            )
+
+        frame_info = dataset.PerFrameFunctionalGroupsSequence[frame_number]
```
Good point. Eugh one-based indexing... 
Addressed in 4bea49d

CPBridge commented on src/highdicom/sr/content.py:
```diff
@@ -523,6 +655,113 @@ def __init__(
+            if hasattr(drv_image, 'SourceImageSequence'):
+                if len(drv_image.SourceImageSequence) != 1:
+                    raise ValueError(
+                        'Could not deduce a single source image from the '
```
Addressed in 4bea49d

CPBridge commented on src/highdicom/sr/content.py:
```diff
@@ -523,6 +655,113 @@ def __init__(
+                    )
+            else:
+                raise AttributeError(
+                    "Dataset does not contain a ReferencedSeriesSequence "
```
Addressed in 4bea49d

CPBridge commented on src/highdicom/sr/content.py:
```diff
@@ -599,3 +838,163 @@ def __init__(
+
+        Notes
+        -----
+             This method will attempt to deduce source image information
```
Nope, probably a vim error...
Addressed in 4bea49d

CPBridge commented:
I have addressed most comments. Will look into checks on the validity of source images (as suggested in the comments) later.

hackermd commented:
> I have addressed most comments.

Could you kindly respond to the comments with `Addressed with #COMMIT_HASH`. This will make it easier for me to keep track of changes and determine whether a comment can be resolved.

CPBridge commented:
Didn't I do that?

hackermd commented:
> Didn't I do that?

You did indeed. Excellent! Sorry I didn't notice. The comments were already resolved and hidden. Please keep comments open and let me close them. That makes it easier for me to keep track of changes.

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -22,6 +24,53 @@
 
 
+def _check_valid_source_image_dataset(dataset: Dataset) -> None:
+    """Raise an error if the image is not a valid for a source image reference.
```
```suggestion
    """Raise an error if the image is not a valid source image reference.
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -103,6 +163,35 @@ def __init__(
 
+    @classmethod
+    def from_source_image(
+            cls,
+            image_dataset: Dataset,
+            referenced_frame_numbers: Optional[Sequence[int]] = None
+        ) -> 'SourceImageForMeasurement':
```
Let's start doing this consistently throughout the library and in compliance with PEP8:

```suggestion
        cls,
        image_dataset: Dataset,
        referenced_frame_numbers: Optional[Sequence[int]] = None
    ) -> 'SourceImageForMeasurement':
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -103,6 +163,35 @@ def __init__(
+
+        Returns
+        -------
+        SourceImageForMeasurement
```
```suggestion
        highdicom.sr.content.SourceImageForMeasurement
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -140,6 +240,35 @@ def __init__(
 
+    @classmethod
+    def from_source_image(
+            cls,
+            image_dataset: Dataset,
+            referenced_frame_numbers: Optional[Sequence[int]] = None
+        ) -> 'SourceImageForRegion':
```
```suggestion
        cls,
        image_dataset: Dataset,
        referenced_frame_numbers: Optional[Sequence[int]] = None
    ) -> 'SourceImageForRegion':
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -140,6 +240,35 @@ def __init__(
+
+        Returns
+        -------
+        SourceImageForRegion
```
```suggestion
        highdicom.sr.content.SourceImageForRegion
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -177,6 +317,35 @@ def __init__(
 
+    @classmethod
+    def from_source_image(
+            cls,
+            image_dataset: Dataset,
+            referenced_frame_numbers: Optional[Sequence[int]] = None
+        ) -> 'SourceImageForSegmentation':
```
```suggestion
        cls,
        image_dataset: Dataset,
        referenced_frame_numbers: Optional[Sequence[int]] = None
    ) -> 'SourceImageForSegmentation':
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -177,6 +317,35 @@ def __init__(
+
+        Returns
+        -------
+        SourceImageForSegmentation
```
```suggestion
        highdicom.sr.content.SourceImageForSegmentation
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -202,6 +371,30 @@ def __init__(self, referenced_series_instance_uid: str):
 
+    @classmethod
+    def from_source_image(
+            cls,
+            image_dataset: Dataset,
+        ) -> 'SourceSeriesForSegmentation':
```
```suggestion
        cls,
        image_dataset: Dataset,
    ) -> 'SourceSeriesForSegmentation':
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -202,6 +371,30 @@ def __init__(self, referenced_series_instance_uid: str):
+
+        Returns
+        -------
+        SourceSeriesForSegmentation
```
```suggestion
        highdicom.sr.content.SourceSeriesForSegmentation
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -414,6 +607,33 @@ def __init__(self, referenced_sop_instance_uid: str):
+
+        Returns
+        -------
+        RealWorldValueMap
```
```suggestion
        highdicom.sr.content.RealWorldValueMap
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -414,6 +607,33 @@ def __init__(self, referenced_sop_instance_uid: str):
 
+    @classmethod
+    def from_source_value_map(
+            cls,
+            value_map_dataset: Dataset,
+        ) -> 'RealWorldValueMap':
```
```suggestion
        cls,
        value_map_dataset: Dataset,
    ) -> 'RealWorldValueMap':
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -524,6 +744,120 @@ def __init__(
 
+    @classmethod
+    def from_seg_dataset(
+            cls,
+            dataset: Dataset,
+            frame_number: int
+        ) -> 'ReferencedSegmentationFrame':
```
```suggestion
        cls,
        dataset: Dataset,
        frame_number: int
    ) -> 'ReferencedSegmentationFrame':
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -524,6 +744,120 @@ def __init__(
+
+        Returns
+        -------
+        ReferencedSegment
```
```suggestion
        highdicom.sr.content.ReferencedSegment
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -524,6 +744,120 @@ def __init__(
+        use the information in the Referenced Series Sequence, which
+        applies to all segments and frames present in the segmentation
+        instance.
+
```
Please add a `Raises` section to indicate under what circumstances a `ValueError` will be raised

CPBridge replied:
Resolved in 4cc5f0e

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -600,3 +934,176 @@ def __init__(
+
+    @classmethod
+    def from_seg_dataset(
+            cls,
+            dataset: Dataset,
+            segment_number: int,
+            frame_numbers: Optional[Sequence[int]] = None
+        ) -> 'ReferencedSegment':
```
```suggestion
        cls,
        dataset: Dataset,
        segment_number: int,
        frame_numbers: Optional[Sequence[int]] = None
    ) -> 'ReferencedSegment':
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -600,3 +934,176 @@ def __init__(
+
+        Returns
+        -------
+        ReferencedSegment
```
```suggestion
        highdicom.sr.content.ReferencedSegment
```

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -103,6 +163,35 @@ def __init__(
+    @classmethod
+    def from_source_image(
+            cls,
+            image_dataset: Dataset,
```
Do you have strong feelings about using `image_dataset` instead of `image`? I generally avoid including the type in the name. We have a type annotation and a docstring to inform the user.

CPBridge replied:
Just that 'image' is very vague and could be a numpy pixel array or something. On balance though it is probably better. Have changed to `image` in this class and all other `SourceXForX` classes, and additionally changed the `dataset` parameter of `ReferencedSegment.from_seg_dataset()` to `segmentation` to match. Changes in 4cc5f0e

On a related note, should we change `from_seg_dataset` to simply `from_seg` or `from_segmentation` to better match `from_source_image`?

hackermd reviewed (changes requested):
Nice! Only a few fixes and suggestions for docstrings.

CPBridge commented:
I have addressed all comments and also added:
- Checks that ensure that referenced frame numbers are valid for the specific dataset passed (previously this was only done for frame numbers that are always incorrect, i.e. < 1)
- Tests for the checks on dataset appropriateness and frame number correctness

hackermd commented on src/highdicom/sr/content.py:
```diff
@@ -938,15 +989,15 @@ def __init__(
     @classmethod
     def from_seg_dataset(
```
Given that we call the related methods `from_source_image()` and `from_source_value_map()`, I wonder whether we should call this method `from_segmentation()` for consistency. What do you think?

CPBridge replied:
I asked you the same question above! I'll make the change

CPBridge replied:
Changed as suggested in 9aca15a @hackermd

hackermd replied:
> I asked you the same question above! I'll make the change

I just realized that, too. We are on the same page 👍

hackermd reviewed (commented):
Thanks @CPBridge! Looks great! I have just one additional comment regarding the name of one of the methods. Please let me know what you think.
